reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java

```diff
@@ -372,89 +375,97 @@ private void drainLoop() {
+									metricsRecorder.recordAllocationFailureAndLatency(clock.millis() - start);
+									poolConfig.allocationStrategy()
+											.returnPermits(1);
+									borrower.fail(error);
+								}
+							}).contextWrite(borrower.currentContext());
 
-						if (permits == 1) {
-							//subscribe to the primary, which will directly feed to the borrower
-							primary.subscribe(alreadyPropagated -> { }, alreadyPropagatedOrLogged -> drain(), this::drain);
-						}
-						else {
-							/*=============================================*
-							 * (warm up in sequence to primary allocation) *
-							 *=============================================*/
-							int toWarmup = permits - 1;
-							logger.debug("should warm up {} extra resources", toWarmup);
-
-							final long startWarmupIteration = clock.millis();
-							// flatMap will eagerly subscribe to the allocator from the current thread, but the concurrency
-							// can be controlled from configuration
-							final int mergeConcurrency = Math.min(poolConfig.allocationStrategy().warmupParallelism(), toWarmup + 1);
-							Flux.range(1, toWarmup)
-									.map(i -> warmupMono(i, toWarmup, startWarmupIteration, allocator).doOnSuccess(__ -> drain()))
-									.startWith(primary.doOnSuccess(__ -> drain()).then())
-									.flatMap(Function.identity(), mergeConcurrency, 1) // since we dont store anything the inner buffer can be simplified
-									.onErrorResume(e -> Mono.empty())
-									.subscribe(aVoid -> { }, alreadyPropagatedOrLogged -> drain(), this::drain);
+							if (permits == 1) {
+								//subscribe to the primary, which will directly feed to the borrower
+								primary.subscribe(alreadyPropagated -> { }, alreadyPropagatedOrLogged -> drain(), this::drain);
+							}
+							else {
+								/*=============================================*
+								 * (warm up in sequence to primary allocation) *
+								 *=============================================*/
+								int toWarmup = permits - 1;
+								logger.debug("should warm up {} extra resources", toWarmup);
+
+								final long startWarmupIteration = clock.millis();
+								// flatMap will eagerly subscribe to the allocator from the current thread, but the concurrency
+								// can be controlled from configuration
+								final int mergeConcurrency = Math.min(poolConfig.allocationStrategy().warmupParallelism(), toWarmup + 1);
+								warmupInProgress = true;
+								Flux.range(1, toWarmup)
+										.map(i -> warmupMono(i, toWarmup, startWarmupIteration, allocator).doOnSuccess(__ -> drain()))
+										.startWith(primary.doOnSuccess(__ -> drain()).then())
+										.flatMap(Function.identity(), mergeConcurrency, 1) // since we dont store anything the inner buffer can be simplified
+										.onErrorResume(e -> Mono.empty())
+										.subscribe(aVoid -> {
+										}, alreadyPropagatedOrLogged -> drain(), () -> {
+											warmupInProgress = false;
+											drain();
+										});
```

Comment:
On error `warmupInProgress` will continue to stay `true`. Is that intensional?

Reply:
good catch, the https://github.com/reactor/reactor-pool/pull/175 is partially addressing this problem.

Reply:
warmpupInProgress should be indeed reset even if there is an error. will think about this.